File: Solutions/2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.dart
class Solution {
  int countPalindromicSubsequence(String s) {
    // Convert the string to a list of characters for easier indexing
    List<String> chars = s.split('');

    // Create a set to store unique 3-letter palindromic subsequences
    Set<String> result = {};

    // Iterate over all lowercase letters from 'a' to 'z'
    for (int i = 0; i < 26; i++) {
      String letter = String.fromCharCode('a'.codeUnitAt(0) + i);

      // Find the first and last occurrence of the current letter
      int first = s.indexOf(letter);
      int last = s.lastIndexOf(letter);

      // If the letter appears at least twice and there is space in between
      if (first != -1 && last != -1 && first < last - 1) {
        // Use a set to collect unique characters between the first and last occurrence
        Set<String> middleChars = {};

        // Iterate through the characters between first and last occurrence
        for (int j = first + 1; j < last; j++) {
          middleChars.add(chars[j]);
        }

        // Each unique middle character forms a unique palindrome of the form: letter + middle + letter
        for (String mid in middleChars) {
          result.add(letter + mid + letter);
        }
      }
    }

    // Return the number of unique palindromic subsequences
    return result.length;
  }
}

File: Solutions/2059-unique-length-3-palindromic-subsequences/unique-length-3-palindromic-subsequences.ts
function countPalindromicSubsequence(s: string): number {

    // Convert string to array for faster indexed access
    const arr = s.split('');

    let total = 0;

    // Loop over all possible characters 'a'..'z'
    for (let code = 97; code <= 122; code++) {
        const c = String.fromCharCode(code);

        // First position of c
        const i = s.indexOf(c);
        // Last position of c
        const j = s.lastIndexOf(c);

        // Must appear at least twice to form c _ c
        if (i !== -1 && i !== j) {

            // Track which middle characters we have seen between i and j
            // A boolean array of size 26, initially false
            const seen = new Array<boolean>(26).fill(false);

            let count = 0;

            // Walk over the interior range (i+1 ... j-1)
            for (let k = i + 1; k < j; k++) {
                const midChar = arr[k];
                const idx = midChar.charCodeAt(0) - 97; // convert 'a'..'z' → 0..25

                // Each distinct middle character produces one palindrome c x c
                if (!seen[idx]) {
                    seen[idx] = true;
                    count++;
                }
            }

            total += count;
        }
    }

    return total;
}
